geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,362 @@
+    final Region region = cache.getRegion(regionName);
+    if (isCancel) {
+      return cancelReplicateRegion(context, region, senderId);
+    }
+
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Sender not running");
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          "Sender is serial and not primary. 0 entries replicated.");
+    }
+
+    try {
+      return executeReplicateFunctionInNewThread(context, region, sender, maxRate, batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution canceled.");
+    } catch (ExecutionException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution failed: " + e.getMessage());
+    }
+  }
+
+  private CliFunctionResult executeReplicateFunctionInNewThread(FunctionContext<String[]> context,
+      Region region, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    Callable<CliFunctionResult> callable =
+        new ReplicateRegionCallable(context, region, sender, maxRate, batchSize);
+    FutureTask<CliFunctionResult> futureTask = new FutureTask<>(callable);
+    Thread t = new Thread(futureTask);
+    String origThreadName = Thread.currentThread().getName();
+    try {
+      t.setName(getReplicateRegionFunctionThreadName(region.getName(),
+          sender.getId()));
+    } finally {
+      Thread.currentThread().setName(origThreadName);
+    }
+    t.start();
+    return futureTask.get();
+  }
+
+  class ReplicateRegionCallable implements Callable<CliFunctionResult> {
+    private final FunctionContext<String[]> context;
+    private final Region region;
+    private final GatewaySender sender;
+    private final long maxRate;
+    private final int batchSize;
+
+    public ReplicateRegionCallable(final FunctionContext<String[]> context, final Region region,
+        final GatewaySender sender, final long maxRate,
+        final int batchSize) {
+      this.context = context;
+      this.region = region;
+      this.sender = sender;
+      this.maxRate = maxRate;
+      this.batchSize = batchSize;
+    }
+
+    @Override
+    public CliFunctionResult call() throws Exception {
+      return replicateRegion(context, region, sender, maxRate, batchSize);
+    }
+  }
+
+  private CliFunctionResult replicateRegion(FunctionContext<String[]> context, Region region,
+      GatewaySender sender, long maxRate, int batchSize) {
+    final Set<?> entries;
+    if (region instanceof PartitionedRegion && sender.isParallel()) {
```

Comment:
I'm not sure it should matter whether the sender is parallel here. You can get the entries the same way either way. If you remove the `sender.isParallel()` check, there won't be any EntrySnapshots in createEvent and I think you'll be able to pass RegionEntry rather than Region.Entry as the parameter all the way down.

Reply:
Oh, right. You're just using the primary sender, so you have to get all the PR entries into the one primary member in the serial case. Thats unfortunate since EntrySnapshots are copies of the RegionEntry, but it looks necessary. Sorry for the confusion.